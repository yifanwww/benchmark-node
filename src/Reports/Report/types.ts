import type { StatisticColumn } from '../../Columns';
import type { Statistics } from '../../Data';

export type StatisticGroup = Statistics[];

export interface BenchmarkResult {
    readonly argLen: number;
    readonly columns: readonly StatisticColumn[];
    readonly paramNames: readonly string[];
    readonly statisticGroups: readonly StatisticGroup[];
}
