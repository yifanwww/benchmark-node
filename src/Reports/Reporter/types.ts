import type { Statistics } from '../../Data';
import type { IIndicator } from '../../Indicators';

export type StatisticGroup = Statistics[];

export interface BenchmarkResult {
    readonly argLen: number;
    readonly indicators: readonly IIndicator[];
    readonly paramNames: readonly string[];
    readonly statisticGroups: readonly StatisticGroup[];
}
