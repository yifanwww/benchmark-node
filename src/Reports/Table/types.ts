export enum ColumnAlign {
    LEFT,
    MEDIUM,
    RIGHT,
}

export type RowID = [number, number];
