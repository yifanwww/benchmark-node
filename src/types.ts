import type { Arguments } from './Parameterization';
import type { UnknownFn } from './types.internal';

/* ----- Types ----- */

/**
 * The type representing millisecond.
 */
export type Millisecond = number & {};

/**
 * The type representing nanosecond.
 *
 * The maximum safe integer in Javascript is `2^53 - 1` (`Number.MAX_SAFE_INTEGER`), can represent nanosecond.
 */
export type Nanosecond = number & {};

export type TestFn = UnknownFn;

export interface BenchmarkingSettings {
    /**
     * The delay between test cycles (ms).
     *
     * Default is `5`, minimum is `1`.
     */
    delay?: Millisecond;
    /**
     * The initial number of ops to run in a benchmark.
     *
     * Default is `16`, minimum is `1`.
     */
    initOps?: number;
    /**
     * The count of measurements required to perform statistical analysis.
     *
     * Default is `15`, minimum is `3`.
     */
    measurementCount?: number;
    /**
     * The minimum time a benchmark uses.
     *
     * Default is `250`, minimum is `1`.
     */
    minMeasurementTime?: Millisecond;
    /**
     * Warm up before formal benchmarking.
     *
     * Default is `7`, minimum is `1`.
     */
    warmupCount?: number;
}

export interface BenchmarkTestFnOptions<T extends TestFn> {
    /**
     * Used for pilot benchmarking and formal benchmarking.
     */
    args?: LooseArray<Arguments<Parameters<T>>>;
    /**
     * Used for jitting benchmarking. The arguments provided in `args` will be also added into `preArgs`.
     */
    jitArgs?: LooseArray<Arguments<Parameters<T>>>;
    /**
     * Set iteration setup.
     * A callback function to be executed exactly once before each benchmark function invocation.
     *
     * Warning: This may change in the future.
     */
    setup?: () => void;
    /**
     * Set iteration cleanup.
     * A callback function to be executed exactly once after each benchmark function invocation.
     *
     * Warning: This may change in the future.
     */
    cleanup?: () => void;
}

/* ----- Type Utils ----- */

export type LooseArray<T> = T | readonly T[];
