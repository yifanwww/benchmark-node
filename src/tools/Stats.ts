import { ConfidenceInterval, _Nanosecond } from '../types.internal';

import { ConsoleLogger } from './ConsoleLogger';
import { Formatter } from './Formatter';
import { MathTool } from './MathTool';
import { Time } from './TimeTool';

/**
 * Class for stats including mean, margin or error, and standard deviation.
 */
export class Stats {
    private _name: string;

    private _n: number;

    private _mean: _Nanosecond;
    private _variance: number;
    private _standardDeviation: _Nanosecond;
    private _standardError: _Nanosecond;
    private _standardErrorPercent: number;

    private _confidenceInterval: ConfidenceInterval;
    private _ciMargin: _Nanosecond;
    private _ciMarginPercent: number;

    private _q0: _Nanosecond;
    private _q1: _Nanosecond;
    private _q2: _Nanosecond;
    private _q3: _Nanosecond;
    private _q4: _Nanosecond;

    private _ops: number;

    public get name() {
        return this._name;
    }

    /**
     * The measurements count.
     */
    public get n() {
        return this._n;
    }

    /**
     * The arithmetic mean of measurements in nanoseconds.
     */
    public get mean() {
        return this._mean;
    }

    /**
     * The variance of measurements.
     */
    public get variance() {
        return this._variance;
    }

    /**
     * The standard deviation of measurements.
     */
    public get standardDeviation() {
        return this._standardDeviation;
    }

    /**
     * The standard error of measurements in nanoseconds.
     */
    public get standardError() {
        return this._standardError;
    }

    /**
     * The margin of error in nanoseconds.
     */
    public get margin() {
        return this._ciMargin;
    }

    public get confidenceInterval() {
        return this._confidenceInterval;
    }

    public get min() {
        return this._q0;
    }

    public get q1() {
        return this._q1;
    }

    public get median() {
        return this._q2;
    }

    public get q3() {
        return this._q3;
    }

    public get max() {
        return this._q4;
    }

    /**
     * The number of executions per second.
     */
    public get ops(): number {
        return this._ops;
    }

    public constructor(name: string, measurements: _Nanosecond[], ops: number) {
        this._name = name;

        measurements.sort((a, b) => a - b);
        measurements = measurements.map((used) => Time.ns(used / ops));

        this._n = measurements.length;

        this._mean = MathTool.mean(measurements);
        this._variance = MathTool.variance(measurements, this._mean);
        this._standardDeviation = Math.sqrt(this._variance) as _Nanosecond;
        this._standardError = (this._standardDeviation / Math.sqrt(this._n)) as _Nanosecond;
        this._standardErrorPercent = (this._standardError / this._mean) * 100;

        const criticalValue = MathTool.getTDistributionCriticalValue(this._n - 1);
        this._ciMargin = (this._standardError * criticalValue) as _Nanosecond;
        this._confidenceInterval = [this._mean - this._ciMargin, this._mean + this._ciMargin] as ConfidenceInterval;
        this._ciMarginPercent = (this._ciMargin / this._mean) * 100;

        this._q0 = measurements[0];
        this._q1 = measurements[Math.round(this._n * 0.25)];
        this._q2 = measurements[Math.round(this._n * 0.5)];
        this._q3 = measurements[Math.round(this._n * 0.75)];
        this._q4 = measurements[this._n - 1];

        this._ops = 1e9 / this._mean;
    }

    public toString(order?: number): string {
        const opsStr = Formatter.beautifyNumber(this._ops.toFixed(this._ops < 100 ? 2 : 0));
        const rmeStr = this._ciMarginPercent.toFixed(2);

        return [
            order === undefined ? '' : `${order}: `,
            `${opsStr} ops/sec`,
            ` ${rmeStr}%`,
            ` (${this._n} sample${this._n > 1 ? 's' : ''})`,
        ].join('');
    }

    public log(): void {
        const logger = ConsoleLogger.default;

        logger.writeLineStatistic(
            [
                `Mean = ${this._mean.toFixed(3)} ns`,
                `StdErr = ${this._standardError.toFixed(3)} ns (${this._standardErrorPercent.toFixed(2)}%)`,
                `N = ${this._n}`,
                `StdDev = ${this._standardDeviation.toFixed(3)} ns`,
            ].join(', '),
        );

        logger.writeLineStatistic(
            [
                `Min = ${this._q0.toFixed(3)} ns`,
                `Q1 = ${this._q1.toFixed(3)} ns`,
                `Median = ${this._q2.toFixed(3)}`,
                `Q3 = ${this._q3.toFixed(3)} ns`,
                `Max = ${this._q4.toFixed(3)} ns`,
            ].join(', '),
        );

        const [left, right] = this._confidenceInterval;
        logger.writeLineStatistic(
            [
                `ConfidenceInterval = [${left.toFixed(3)} ns; ${right.toFixed(3)} ns] (CI 95%)`,
                `Margin = ${this._ciMargin.toFixed(3)} ns (${this._ciMarginPercent.toFixed(2)}% of Mean)`,
            ].join(', '),
        );

        logger.writeLine();
    }
}
