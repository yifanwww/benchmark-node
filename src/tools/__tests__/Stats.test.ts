import { _Nanosecond } from '../../types.internal';
import { Stats } from '../Stats';

describe(`test class \`${Stats.name}\``, () => {
    it('calculates', () => {
        const stats = new Stats(
            'name',
            [
                37051080, 33390980, 33920080, 33423180, 33684980, 34514380, 33498580, 34333480, 33923280, 34479480,
                33398680, 33242780, 33088980, 33790080, 33530180,
            ] as _Nanosecond[],
            141386309,
        );

        expect(JSON.stringify(stats, undefined, 4)).toMatchSnapshot();
    });
});
