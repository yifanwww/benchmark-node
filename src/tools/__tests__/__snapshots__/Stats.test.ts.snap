// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test class \`Stats\` calculates 1`] = `
"{
    \\"_name\\": \\"name\\",
    \\"_n\\": 15,
    \\"_mean\\": 0.2401317843771328,
    \\"_variance\\": 0.00004636426559949729,
    \\"_standardDeviation\\": 0.006809131045845519,
    \\"_standardError\\": 0.001758110076180239,
    \\"_standardErrorPercent\\": 0.7321438437400208,
    \\"_ciMargin\\": 0.0037711461134066126,
    \\"_confidenceInterval\\": [
        0.2363606382637262,
        0.24390293049053943
    ],
    \\"_ciMarginPercent\\": 1.5704485448223444,
    \\"_q0\\": 0.2340324196453845,
    \\"_q1\\": 0.2363961562926153,
    \\"_q2\\": 0.23899117417373134,
    \\"_q3\\": 0.24283454489217907,
    \\"_q4\\": 0.26205564217678246,
    \\"_ops\\": 4164379999.0653286
}"
`;
